File: src/main/java/com/libreshockwave/chunks/TextChunk.java
package com.libreshockwave.chunks;

import com.libreshockwave.format.ChunkType;
import com.libreshockwave.io.BinaryReader;

import java.util.ArrayList;
import java.util.List;

/**
 * Styled text chunk (STXT).
 * Contains text content with formatting information.
 */
public record TextChunk(
    int id,
    String text,
    List<TextRun> runs
) implements Chunk {

    @Override
    public ChunkType type() {
        return ChunkType.STXT;
    }

    public record TextRun(
        int startOffset,
        int endOffset,
        int fontId,
        int fontSize,
        int fontStyle
    ) {}

    public static TextChunk read(BinaryReader reader, int id) {
        int headerLen = reader.readI32();
        int textLen = reader.readI32();

        reader.skip(headerLen - 8);

        String text = reader.readStringMacRoman(textLen);

        // Parse formatting runs if present
        List<TextRun> runs = new ArrayList<>();
        if (reader.bytesLeft() >= 4) {
            int runCount = reader.readI16();
            reader.skip(2);

            for (int i = 0; i < runCount && reader.bytesLeft() >= 16; i++) {
                int startOffset = reader.readI32();
                int fontId = reader.readI16();
                int fontStyle = reader.readU8();
                reader.skip(1);
                int fontSize = reader.readI16();
                reader.skip(6);

                runs.add(new TextRun(startOffset, textLen, fontId, fontSize, fontStyle));
            }
        }

        return new TextChunk(id, text, runs);
    }
}

File: cpp/src/chunks/TextChunk.cpp
#include "libreshockwave/chunks/TextChunk.hpp"

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::chunks {

TextChunk::TextChunk(const DirectorFile* file, id::ChunkId id, std::string text, std::vector<TextRun> runs)
    : file_(file), id_(id), text_(std::move(text)), runs_(std::move(runs)) {}

const DirectorFile* TextChunk::file() const { return file_; }
format::ChunkType TextChunk::type() const { return format::ChunkType::STXT; }
id::ChunkId TextChunk::id() const { return id_; }
const std::string& TextChunk::text() const { return text_; }
const std::vector<TextChunk::TextRun>& TextChunk::runs() const { return runs_; }

TextChunk TextChunk::read(const DirectorFile* file, io::BinaryReader& reader, id::ChunkId id, int directorVersion) {
    const auto originalOrder = reader.order();
    reader.setOrder(io::ByteOrder::BigEndian);

    const int headerLen = reader.readI32();
    const int textLen = reader.readI32();
    if (headerLen > 8) {
        reader.skip(static_cast<std::size_t>(headerLen - 8));
    }

    auto text = reader.readStringMacRoman(static_cast<std::size_t>(textLen));
    std::vector<TextRun> runs;

    if (reader.bytesLeft() >= 4) {
        const int runCount = reader.readI16();
        reader.skip(2);

        for (int index = 0; index < runCount && reader.bytesLeft() >= 16; ++index) {
            const int startOffset = reader.readI32();
            const int fontId = reader.readI16();
            const int fontStyle = reader.readU8();
            int fontSize = 0;
            if (usesLegacyRunLayout(directorVersion)) {
                reader.skip(3);
                fontSize = reader.readI16();
            } else {
                reader.skip(1);
                fontSize = reader.readI16();
                reader.skip(2);
            }
            const int colorR = reader.readU8();
            const int colorG = reader.readU8();
            const int colorB = reader.readU8();
            reader.skip(1);

            runs.push_back(TextRun{startOffset, textLen, fontId, fontSize, fontStyle, colorR, colorG, colorB});
        }
    }

    reader.setOrder(originalOrder);
    return TextChunk(file, id, std::move(text), std::move(runs));
}

bool TextChunk::usesLegacyRunLayout(int directorVersion) {
    return directorVersion > 0 && directorVersion <= 1600;
}

} // namespace libreshockwave::chunks
